File: app/src/main/java/ua/meugen/android/pdfcomposer/app/di/Singleton.java
package ua.meugen.android.pdfcomposer.app.di;

import java.lang.ref.SoftReference;

/**
 * @author meugen
 */
abstract class Singleton<T> {

    static <T> Singleton<T> create(final boolean soft) {
        return soft
                ? new SoftSingleton<>()
                : new StrongSingleton<>();
    }

    abstract T getSingleton();

    abstract void putSingleton(final T inst);

    T get(final Creator<T> creator) {
        T result = getSingleton();
        if (result == null) {
            result = creator.create();
            putSingleton(result);
        }
        return result;
    }

    private static class SoftSingleton<T> extends Singleton<T> {

        private SoftReference<T> ref;

        @Override
        T getSingleton() {
            return ref == null ? null : ref.get();
        }

        @Override
        void putSingleton(final T inst) {
            ref = new SoftReference<>(inst);
        }
    }

    private static class StrongSingleton<T> extends Singleton<T> {

        private T ref;

        @Override
        T getSingleton() {
            return ref;
        }

        @Override
        void putSingleton(final T inst) {
            ref = inst;
        }
    }

    interface Creator<T> {

        T create();
    }
}

File: app/src/main/java/ua/meugen/android/pdfcomposer/app/di/Singleton.kt
package ua.meugen.android.pdfcomposer.app.di

import java.lang.ref.SoftReference

/**
 * @author meugen
 */
abstract class Singleton<T> {

    abstract val singleton: T?

    abstract fun putSingleton(inst: T)

    operator fun get(creator: Creator<T>): T {
        singleton?.let { return it }
        val instance = creator.create()
        putSingleton(instance)
        return instance
    }

    private class SoftSingleton<T> : Singleton<T>() {

        private var ref: SoftReference<T>? = null

        override val singleton: T?
            get() = ref?.get()

        override fun putSingleton(inst: T) {
            ref = SoftReference<T>(inst)
        }
    }

    private class StrongSingleton<T> : Singleton<T>() {

        override var singleton: T? = null
            private set

        override fun putSingleton(inst: T) {
            singleton = inst
        }
    }

    interface Creator<T> {

        fun create(): T
    }

    companion object {

        fun <T> create(soft: Boolean): Singleton<T>
                = if (soft) SoftSingleton() else StrongSingleton()
    }
}
